src/Normalizer/EntityHelperTrait.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace Drupal\fb_instant_articles\Normalizer;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Entity\EntityChangedInterface;
+use Drupal\user\EntityOwnerInterface;
+
+/**
+ * Common entity data getters shared between the normalizers.
+ */
+trait EntityHelperTrait {
+
+  /**
+   * Instant articles config.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  protected $config;
+
+  /**
+   * Helper function to compute the canonical URL for a given entity.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   Content entity being normalized.
+   *
+   * @return string
+   *   The canonical URL for the given entity.
+   *
+   * @throws \Drupal\Core\Entity\EntityMalformedException
+   */
+  public function entityCanonicalUrl(ContentEntityInterface $entity) {
+    if ($override = $this->config->get('canonical_url_override')) {
```

Comment:
we could add linkTemplate to entities via [hook_entity_type_alter](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Entity%21entity.api.php/function/hook_entity_type_alter/8.5.x) instead of keeping at config. Something like [this](http://cgit.drupalcode.org/drupal/tree/core/modules/settings_tray/settings_tray.module?h=8.5.x#n99)

Reply:
Oh that's interesting, TIL. Let me file a follow up issue for that. I think that could be worthwhile, but don't want to get too distracted at the moment.

Reply:
Follow up issue: https://www.drupal.org/project/fb_instant_articles/issues/2994704